contracts/interchain-token/tests/test.rs

```diff
@@ -4,55 +4,515 @@ extern crate std;
+
+    assert_invoke_auth_ok!(
+        user3,
+        token.try_transfer_from(&user3, &user2, &user1, &400_i128)
+    );
+    assert_eq!(
+        env.auths(),
+        std::vec![(
+            user3.clone(),
+            AuthorizedInvocation {
+                function: AuthorizedFunction::Contract((
+                    token.address.clone(),
+                    Symbol::new(&env, "transfer_from"),
+                    (&user3, &user2, &user1, 400_i128).into_val(&env),
+                )),
+                sub_invocations: std::vec![]
+            }
+        )]
+    );
+    assert_eq!(token.balance(&user1), 800_i128);
+    assert_eq!(token.balance(&user2), 200_i128);
+
+    assert_invoke_auth_ok!(user1, token.try_transfer(&user1, &user3, &300_i128));
+    assert_eq!(token.balance(&user1), 500_i128);
+    assert_eq!(token.balance(&user3), 300_i128);
+
+    assert_invoke_auth_ok!(owner1, token.try_transfer_ownership(&owner2));
+    assert_eq!(
+        env.auths(),
+        std::vec![(
+            owner1.clone(),
+            AuthorizedInvocation {
+                function: AuthorizedFunction::Contract((
+                    token.address.clone(),
+                    Symbol::new(&env, "transfer_ownership"),
+                    (&owner2,).into_val(&env),
+                )),
+                sub_invocations: std::vec![]
+            }
+        )]
+    );
+
+    // Increase to 500
+    assert_invoke_auth_ok!(
+        user2,
+        token.try_approve(&user2, &user3, &500_i128, &expiration_ledger)
+    );
+    assert_eq!(token.allowance(&user2, &user3), 500_i128);
+
+    assert_invoke_auth_ok!(
+        user2,
+        token.try_approve(&user2, &user3, &0_i128, &expiration_ledger)
+    );
+    assert_eq!(
+        env.auths(),
+        std::vec![(
+            user2.clone(),
+            AuthorizedInvocation {
+                function: AuthorizedFunction::Contract((
+                    token.address.clone(),
+                    symbol_short!("approve"),
+                    (&user2, &user3, 0_i128, expiration_ledger).into_val(&env),
+                )),
+                sub_invocations: std::vec![]
+            }
+        )]
+    );
+    assert_eq!(token.allowance(&user2, &user3), 0);
+}
 
-fn setup_env<'a>() -> (Env, InterchainTokenClient<'a>) {
+#[test]
+fn minter_test() {
     let env = Env::default();
+
+    let amount = 1000;
     let owner = Address::generate(&env);
-    let contract_id = env.register(InterchainToken, (&owner,));
-    let client = InterchainTokenClient::new(&env, &contract_id);
+    let minter1 = Address::generate(&env);
+    let user = Address::generate(&env);
 
-    (env, client)
+    let token = create_token(&env, &owner, &minter1);
+
+    assert_invoke_auth_err!(owner, token.try_mint(&minter1, &user, &amount));
+    assert_invoke_auth_err!(user, token.try_mint(&minter1, &user, &amount));
+    assert_invoke_auth_ok!(minter1, token.try_mint(&minter1, &user, &amount));
```

Comment:
this should be two separate tests, something like `mint_from_minter_succeeds` and `mint_from_invalid_minter_fails`

Reply:
updated.